File: src/java/engine/movegen/Location.java
package java.engine.movegen;

import java.engine.board.Board;

import static java.engine.board.Board.SIZE;

/**
 * Created by melvic on 7/3/18.
 */
public interface Location {
    public enum File { A, B, C, D, E, F, G, H }
    public enum Rank { _1, _2, _3, _4, _5, _6, _7, _8 }

    public File getFile();
    public Rank getRank();

    public static Location of(File file, Rank rank) {
        return new Location() {
            @Override
            public File getFile() {
                return file;
            }

            @Override
            public Rank getRank() {
                return rank;
            }

            @Override
            public boolean equals(Object obj) {
                if (!(obj instanceof Location)) return false;
                Location location = (Location) obj;
                return location.getFile().equals(getFile())
                        && location.getRank().equals(getRank());
            }

            @Override
            public String toString() {
                return getFile().toString() + getRank().toString();
            }
        };
    }

    public static Location of(int row, int col) {
        File file = File.values()[col];
        Rank rank = Rank.values()[SIZE - 1 - row];
        return Location.of(file, rank);
    }

    public static Location of(int position) {
        return of(position / Board.SIZE, position % Board.SIZE);
    }
}

File: src/engine/movegen/Location.scala
package engine.movegen

import engine.utils.Implicits.Locations._

/**
  * Created by melvic on 8/6/18.
  */
sealed trait File
case object A extends File
case object B extends File
case object C extends File
case object D extends File
case object E extends File
case object F extends File
case object G extends File
case object H extends File

sealed trait Rank
case object _1 extends Rank
case object _2 extends Rank
case object _3 extends Rank
case object _4 extends Rank
case object _5 extends Rank
case object _6 extends Rank
case object _7 extends Rank
case object _8 extends Rank

case class Location(file: File, rank: Rank)

object Location {
  lazy val Files: List[File] = A :: B :: C :: D :: E :: F :: G :: H :: Nil
  lazy val Ranks: List[Rank] = _1 :: _2 :: _3 :: _4 :: _5 :: _6 :: _7 :: _8 :: Nil
}

